### PR TITLE
Fix thread-safety race condition in EnsureTaskHubWorker

### LINKED ISSUE #3343: Multiple various exceptions thrown when starting isolated durable function on Consumption Plan after upgrading from v1.13.1
### Description
After upgrading from v1.13.1 to 1.14,*, we have observed multiple instances of `System.InvalidOperationException` on attempted startup of our durable functions when running on Consumption Plan in Azure. Example exception messages include:

- The listener for function 'FunctionNameWithCosmosDBTrigger' was unable to start. This builder instance has already been used to build a processor. Create a new instance to build another. 
- Exception while executing function: FunctionNameWithTimerTrigger. The local RPC address has not been configured!
- The listener for function 'FunctionNameOrchestrationFunction' was unable to start. The orchestration service has already started.
 
### Expected behavior
All functions and listeners should start without exceptions thrown

### Actual behavior
Multiple exceptions are thrown, often resulting in the durable function not starting at all

### Known workarounds
Revert to v1.13.1 of the package

### App Details

- **Durable Functions extension versions**: 1.14.0 and 1.14.1
- **Azure Functions runtime version**: 4.1046.100.25616
- **Programming language used**: C# .NET 10
- **App Service Plan**:  Consumption Plan, Windows

### If deployed to Azure
No orchestrations are running as the functions fail to start

### PR DESCRIPTION
**Pull request by torosent**
# Summary
## What changed?
- Made `EnsureTaskHubWorker()` thread-safe by adding a double-check locking pattern with a dedicated `taskHubWorkerInitLock` object. The previous `??=` operator is not atomic and allowed concurrent threads to each create a separate `TaskHubWorker` instance.
- Made `AspNetCoreLocalGrpcListener.StartAsync()` thread-safe using a `SemaphoreSlim`-based double-check locking pattern. Core startup logic extracted into `StartInternalAsync()`.
- Added concurrency unit tests validating both fixes.

## Why is this change needed?
- On Consumption Plan cold starts with multiple triggers (e.g., Timer + CosmosDB + Orchestration), `GetClient()` (called without `taskHubLock`) and `StartTaskHubWorkerIfNotStartedAsync()` (called with `taskHubLock`) can race into `EnsureTaskHubWorker()` simultaneously. Since `??=` is not atomic, two threads can both see `this.taskHubWorker == null`, both call `InitializeTaskHubWorker()`, and the second one overwrites the first. This causes a cascade of errors:
  1. **`NullReferenceException`** — The overwritten `TaskHubWorker` (never started) has null dispatchers
  2. **"The orchestration service has already started"** — The replacement `TaskHubWorker` tries to start the shared `AzureStorageOrchestrationService` that was already started by the first
  3. **"The local RPC address has not been configured"** — gRPC listener state is lost when the second initialization overwrites the first

- This regression was introduced in v3.10.0 by PR #3260 which extracted `TaskHubWorker` creation from `StartTaskHubWorkerIfNotStartedAsync()` (always under `taskHubLock`) into the new `EnsureTaskHubWorker()` method (called from `GetClient()` without any lock).

## Issues / work items
- Resolves #3343

---

# Project checklist
- [x] Documentation changes are not required
- [x] Release notes are not required for the next release
- [ ] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [x] No extra work is required to be leveraged by OutOfProc SDKs
- [x] No change to the version of the `WebJobs.Extensions.DurableTask` package
- [x] No EventIds were added to `EventSource` logs
- [ ] This change should be added to the `v2.x` branch
  - [x] Otherwise: This change applies exclusively to `WebJobs.Extensions.DurableTask` v3.x and will be retained only in the `dev` and `main` branches
- [x] Breaking change? No

---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [x] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [ ] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot (Claude Opus 4.6)
- AI-assisted areas/files:
  - `src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs` — double-check locking in `EnsureTaskHubWorker()`
  - `src/WebJobs.Extensions.DurableTask/Grpc/AspNetCoreLocalGrpcListener.cs` — `SemaphoreSlim`-based double-check locking in `StartAsync()`
  - `test/FunctionsV2/DurableTaskExtensionConcurrencyTests.cs` — concurrency unit tests
- What you changed after AI output: Reviewed and validated the fix against the race condition analysis

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing
## Automated tests
- Result: Passed — 285 passed (including 2 new concurrency tests), 389 pre-existing failures (DI/config issues unrelated to this change, identical count on `dev` baseline)
- New tests:
  - `EnsureTaskHubWorker_ConcurrentCalls_ReturnsSameInstance` — verifies 10 threads calling `EnsureTaskHubWorker()` simultaneously all get the same instance
  - `AspNetCoreLocalGrpcListener_ConcurrentStartAsync_StartsOnlyOnce` — verifies 10 concurrent `StartAsync()` calls result in a single listener

## Manual validation (only if runtime/behavior changed)
- Not yet performed. Recommend testing on a Consumption Plan app with Timer + CosmosDB + Orchestration triggers to confirm the race condition no longer occurs on cold start.

---

# Notes for reviewers
- The race condition only manifests on Consumption Plan cold starts where multiple non-orchestration triggers (Timer, CosmosDB, etc.) bind `DurableTaskClient` concurrently with the orchestration listener starting up.
- The fix uses the standard double-check locking pattern: a fast-path null check avoids the lock on the hot path, while the lock + `??=` inside ensures only one thread creates the `TaskHubWorker`.
- The `AspNetCoreLocalGrpcListener.StartAsync()` now uses `SemaphoreSlim` (async-compatible) with the same double-check pattern, extracting the core logic into `StartInternalAsync()`.

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs
Comment (bachuv) [inline src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:1]: Should we add a test for these changes?

Comment (torosent) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:1]: Fixed

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs [hunk 1]
```diff
@@ -346,7 +347,15 @@ internal string GetBackendInfo()
...
-            return this.taskHubWorker ??= this.InitializeTaskHubWorker();
+            if (this.taskHubWorker != null)
+            {
+                return this.taskHubWorker;
+            }
+
+            lock (this.taskHubWorkerInitLock)
+            {
+                return this.taskHubWorker ??= this.InitializeTaskHubWorker();
+            }
```
[the remark this answers is not in the document]
Comment (torosent) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:358]: Fixed

File: src/WebJobs.Extensions.DurableTask/Grpc/AspNetCoreLocalGrpcListener.cs [hunk 2]
```diff
@@ -33,6 +33,12 @@ public AspNetCoreLocalGrpcListener(DurableTaskExtension extension)
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
+            // Prevent double-start: if already started, return immediately
+            if (this.host != null)
+            {
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (torosent) [reply src/WebJobs.Extensions.DurableTask/Grpc/AspNetCoreLocalGrpcListener.cs]: Fixed

File: test/FunctionsV2/DurableTaskExtensionConcurrencyTests.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+            await Task.WhenAll(tasks);
+
+            // Assert
+            Assert.NotNull(listener.ListenAddress);
+            Assert.True(Uri.TryCreate(listener.ListenAddress, UriKind.Absolute, out Uri uri));
+            Assert.True(uri.IsLoopback);
+
+            // Cleanup
+            await listener.StopAsync(default);
```
[the remark this answers is not in the document]
Comment (torosent) [reply test/FunctionsV2/DurableTaskExtensionConcurrencyTests.cs:91]: Already handled. The test calls listener.StopAsync(default) at the end, which stops the Kestrel host. Without IDisposable on the class, there's nothing additional to dispose.

File: test/FunctionsV2/DurableTaskExtensionConcurrencyTests.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        }
```
[the remark this answers is not in the document]
Comment (torosent) [reply test/FunctionsV2/DurableTaskExtensionConcurrencyTests.cs:92]: A single valid ListenAddress confirms only one host was started (each call would bind a different port). Adding a call counter would require modifying production code for testability

File: src/WebJobs.Extensions.DurableTask/Grpc/AspNetCoreLocalGrpcListener.cs [hunk 5]
```diff
@@ -22,6 +22,7 @@ internal class AspNetCoreLocalGrpcListener : ILocalGrpcListener
     {
         private const string HostName = "127.0.0.1";
         private readonly DurableTaskExtension extension;
+        private readonly SemaphoreSlim startLock = new SemaphoreSlim(1, 1);
```
[the remark this answers is not in the document]
Comment (torosent) [reply src/WebJobs.Extensions.DurableTask/Grpc/AspNetCoreLocalGrpcListener.cs:25]: AspNetCoreLocalGrpcListener lives for the entire process lifetime and is never explicitly disposed. SemaphoreSlim's unmanaged resources are minimal and will be cleaned up by GC. Adding IDisposable would also change the class's contract (ripple to ILocalGrpcListener, DI registration, etc.)

File: src/WebJobs.Extensions.DurableTask/Grpc/AspNetCoreLocalGrpcListener.cs [hunk 6]
```diff
@@ -32,6 +33,29 @@ public AspNetCoreLocalGrpcListener(DurableTaskExtension extension)
         public string? ListenAddress { get; private set; }
 
         public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            // Fast check: if already started, return immediately
+            if (this.host != null)
+            {
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (torosent) [reply src/WebJobs.Extensions.DurableTask/Grpc/AspNetCoreLocalGrpcListener.cs:41]: Fixed. `StartInternalAsync` now uses a local `IHost newHost` variable and only assigns `this.host = newHost` as the very last statement — after `host.StartAsync()`, `ListenAddress` assignment, and logging are all complete. This ensures a concurrent caller hitting the fast-path `if (this.host != null)` check will never return before initialization is fully done.

As a bonus, if `StartInternalAsync` throws partway through, `this.host` stays null, so a retry is possible.

File: src/WebJobs.Extensions.DurableTask/Grpc/AspNetCoreLocalGrpcListener.cs [hunk 7]
```diff
@@ -32,6 +33,29 @@ public AspNetCoreLocalGrpcListener(DurableTaskExtension extension)
...
+
+            await this.startLock.WaitAsync(cancellationToken);
+            try
+            {
+                // Double-check after acquiring the lock
+                if (this.host == null)
+                {
+                    await this.StartInternalAsync(cancellationToken);
+                }
```
[the remark this answers is not in the document]
Comment (torosent) [reply src/WebJobs.Extensions.DurableTask/Grpc/AspNetCoreLocalGrpcListener.cs:50]: Fixed. `StartInternalAsync` now uses a local `IHost newHost` variable and only assigns `this.host = newHost` as the very last statement, after full initialization. If `StartInternalAsync` throws partway through, `this.host` stays null so a retry is possible.

File: test/FunctionsV2/DurableTaskExtensionConcurrencyTests.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            // Act
+            var tasks = Enumerable.Range(0, taskCount)
+                .Select(_ => listener.StartAsync(default))
+                .ToArray();
+
+            await Task.WhenAll(tasks);
+
+            // Assert
+            Assert.NotNull(listener.ListenAddress);
+            Assert.True(Uri.TryCreate(listener.ListenAddress, UriKind.Absolute, out Uri uri));
```
[the remark this answers is not in the document]
Comment (torosent) [reply test/FunctionsV2/DurableTaskExtensionConcurrencyTests.cs]: Fixed. Test now collects `ListenAddress` per-task into a `ConcurrentBag` and asserts all entries are valid, ensuring each concurrent caller sees a valid address after its own `StartAsync` returns.

**Review by bachuv — APPROVED** [review]
